### PR TITLE
Add AVX and FMA intrinsics in Factorization Machine

### PR DESCRIPTION
**Pull request by pkumar07**
Added AVX and FMA C++ intrinsics in factorizationmachinenative.dll which currently implements C++ SSE code as suggested in #3000.

### CODE REVIEW

File: src/Microsoft.ML.StandardTrainers/FactorizationMachine/FactorizationMachineInterface.netcoreapp.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
+using System.Runtime.CompilerServices;
```
Comment (wschin) [inline src/Microsoft.ML.StandardTrainers/FactorizationMachine/FactorizationMachineInterface.netcoreapp.cs]: Missing license.

File: src/Microsoft.ML.StandardTrainers/FactorizationMachine/FactorizationMachineInterface.netstandard.cs [hunk 2]
```diff
@@ -30,11 +30,11 @@ private static bool Compat(AlignedArray a)
         }
 
         [DllImport(NativePath), SuppressUnmanagedCodeSecurity]
-        public static extern void CalculateIntermediateVariablesNative(int fieldCount, int latentDim, int count, int* /*const*/ fieldIndices, int* /*const*/ featureIndices,
+        public static extern void CalculateIntermediateVariablesNativeSSE(int fieldCount, int latentDim, int count, int* /*const*/ fieldIndices, int* /*const*/ featureIndices,
```
Comment (wschin) [inline src/Microsoft.ML.StandardTrainers/FactorizationMachine/FactorizationMachineInterface.netstandard.cs]: CalculateIntermediateVariablesNativeSSE ---> CalculateIntermediateVariablesSSE

File: src/Native/FactorizationMachineNative/FactorizationMachineCoreAVX.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    const int * pi = featureIndices;
+    const float * px = featureValues;
+    const float * pq = latentSum;
+    float * pw = linearWeights;
+    float * pv = latentWeights;
+    float * phw = linearAccumulatedSquaredGrads;
+    float * phv = latentAccumulatedSquaredGrads;
+
+    const __m256 _wei = _mm256_set1_ps(weight);
+    const __m256 _s = _mm256_set1_ps(slope);
```
Comment (wschin) [inline src/Native/FactorizationMachineNative/FactorizationMachineCoreAVX.cpp]: Does our memory alignment meet AVX's requirement?

File: src/Native/FactorizationMachineNative/FactorizationMachineCoreSSE.cpp [hunk 4]
```diff
@@ -8,10 +8,10 @@
 #include <limits>
 #include <pmmintrin.h>
 
-// Compute the output value of the field-aware factorization, as the sum of the linear part and the latent part. 
+// Compute the output value of the field-aware factorization, as the sum of the linear part and the latent part.
```
Comment (wschin) [inline src/Native/FactorizationMachineNative/FactorizationMachineCoreSSE.cpp]: Given the difficulty of reading AVX code, please add a comment:
// This function implements Algorithm 1 in [https://github.com/wschin/fast-ffm/blob/master/fast-ffm.pdf](https://github.com/wschin/fast-ffm/blob/master/fast-ffm.pdf)
.

Please add the same line to SSE and FMA counterparts.

File: src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp [hunk 5]
```diff
@@ -105,9 +105,9 @@ EXPORT_API(void) CalculateIntermediateVariablesNative(int fieldCount, int latent
     *response = linearResponse + latentResponse;
 }
 
-// Calculate the stochastic gradient and update the model. 
+// Calculate the stochastic gradient and update the model.
 // The /*const*/ comment on the parameters of the function means that their values should not get altered by this function.
-EXPORT_API(void) CalculateGradientAndUpdateNative(float lambdaLinear, float lambdaLatent, float learningRate, int fieldCount, int latentDim, float weight, int count,
```
Comment (wschin) [inline src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp]: Given the difficulty of reading AVX code, please add a comment:
// This function implements Algorithm 2 in [https://github.com/wschin/fast-ffm/blob/master/fast-ffm.pdf](https://github.com/wschin/fast-ffm/blob/master/fast-ffm.pdf)
.

Please add the same line to SSE and FMA counterparts.

File: src/Native/FactorizationMachineNative/FactorizationMachineCoreFMA.cpp [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                __m256 _q = _mm256_load_ps(qfprimef + k);
+
+                // Calculate L2-norm regularization's gradient.
+                __m256 _g = _mm256_mul_ps(_lambdav, _v);
+
+                // Calculate loss function's gradient.
+                if (fprime != f)
+                   _g = _mm256_fmadd_ps(_sx, _q, _g);
+                else
+                   _g = _mm256_fmadd_ps(_sx, _mm256_sub_ps(_q, _mm256_mul_ps(_v, _x)), _g);
```
Comment (wschin) [inline src/Native/FactorizationMachineNative/FactorizationMachineCoreFMA.cpp]: Not sure if we can avoid this potential branch. As least, I failed a long time ago. :)

File: build/Dependencies.props [hunk 7]
```diff
@@ -7,6 +7,7 @@
     <SystemCollectionsImmutableVersion>1.5.0</SystemCollectionsImmutableVersion>
     <SystemMemoryVersion>4.5.1</SystemMemoryVersion>
     <SystemReflectionEmitLightweightPackageVersion>4.3.0</SystemReflectionEmitLightweightPackageVersion>
+    <SystemRuntimeCompilerServices>4.5.2</SystemRuntimeCompilerServices>
     <SystemThreadingTasksDataflowPackageVersion>4.8.0</SystemThreadingTasksDataflowPackageVersion>
```
Comment (wschin) [inline build/Dependencies.props]: Any reason we need this line? Could you add a comment to add some details?

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 8]
```diff
@@ -9,6 +9,7 @@
 
 namespace Microsoft.ML.Internal.CpuMath
 {
+    [BestFriend]
```
Comment (wschin) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: >BestFriend [](start = 5, length = 10)

Nice!

File: src/Microsoft.ML.StandardTrainers/Microsoft.ML.StandardTrainers.csproj [hunk 9]
```diff
@@ -1,7 +1,8 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework Condition="'$(UseIntrinsics)' != 'true'">netstandard2.0</TargetFramework>
```
Comment (wschin) [inline src/Microsoft.ML.StandardTrainers/Microsoft.ML.StandardTrainers.csproj]: @ericstj, could you please take a look? I am not quite familiar with the build settings. Thank you!

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs (hunk 8 above)
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: Why is this change necessary?

File: src/Native/FactorizationMachineNative/CMakeLists.txt [hunk 10]
```diff
@@ -1,12 +1,16 @@
 project (FactorizationMachineNative)
 
 set(SOURCES
-    FactorizationMachineCore.cpp
+    FactorizationMachineCoreSSE.cpp
```
Comment (wschin) [inline src/Native/FactorizationMachineNative/CMakeLists.txt]: Are we compiling all of source files? Are all of them cross-platform?

File: src/Native/FactorizationMachineNative/FactorizationMachineCoreAVX.cpp [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "../Stdafx.h"
+#include <cmath>
+#include <cstring>
+#include <limits>
+#include <immintrin.h>
+//check loads
+EXPORT_API(void) CalculateIntermediateVariablesNativeAVX(int fieldCount, int latentDim, int count, _In_ int * fieldIndices, _In_ int * featureIndices, _In_ float * featureValues,
```
Comment (eerhardt) [inline src/Native/FactorizationMachineNative/FactorizationMachineCoreAVX.cpp]: Why write these functions in C++ when they are only called from netcoreapp3.0? Why not write them in C# like the rest of our intrinsics in netcoreapp3.0?

File: src/Microsoft.ML.StandardTrainers/Microsoft.ML.StandardTrainers.csproj [hunk 12]
```diff
@@ -1,7 +1,8 @@
...
+    <TargetFrameworks Condition="'$(UseIntrinsics)' == 'true'">netstandard2.0;netcoreapp3.0</TargetFrameworks>
```
Comment (eerhardt) [inline src/Microsoft.ML.StandardTrainers/Microsoft.ML.StandardTrainers.csproj]: The NuGet package is not going to work the way this is currently written. See [https://github.com/dotnet/machinelearning/issues/534](https://github.com/dotnet/machinelearning/issues/534) for more information. Basically, once you introduce a single assembly in `lib\netcoreapp3.0` of the NuGet package, ALL the assemblies need to be in the `lib\netcoreapp3.0` folder of the package.

It would probably be easiest if the new AVX/FMA were put in the CpuMath assembly/package instead.

File: build/Dependencies.props (hunk 7 above)
Comment (pkumar07) [reply build/Dependencies.props]: Hi Wei-Sheng! Thanks for reviewing. It’s not needed for now and I will remove it.

File: src/Native/FactorizationMachineNative/CMakeLists.txt (hunk 10 above)
Comment (pkumar07) [reply src/Native/FactorizationMachineNative/CMakeLists.txt]: Yes, we are compiling all the source files. Not all of them are cross-platform. Avx.isSupported (and similar) dispatches the code as needed.

File: src/Native/FactorizationMachineNative/FactorizationMachineCoreAVX.cpp (hunk 3 above)
Comment (pkumar07) [reply src/Native/FactorizationMachineNative/FactorizationMachineCoreAVX.cpp]: Thanks for pointing it out. I will take a look at the memory alignment for AVX/FMA.

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs (hunk 8 above)
Comment (pkumar07) [reply src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: Hi Eric! Access to CpuMathUtils from FactorizationMachineInterface gives an error when using netcoreapp3.0. Issue #3654

File: src/Native/FactorizationMachineNative/FactorizationMachineCoreAVX.cpp (hunk 11 above)
Comment (pkumar07) [reply src/Native/FactorizationMachineNative/FactorizationMachineCoreAVX.cpp]: That’s a good point. I will try implementing the same in C# and will close this PR for now.

**Comment by wschin** [comment]
@pkumar07, why did you close it?

**Comment by pkumar07** [comment]
@wschin 
As @eerhardt mentioned that these native functions are only being called from netcoreapp3.0, it would be better to write them in C#. I can try implementing these functions in C#. What do you suggest?
